File: Editor/src/main/java/org/turbomedia/turboedit/editor/panes/file_browser/FileBrowserPane.java
package org.turbomedia.turboedit.editor.panes.file_browser;

import javafx.geometry.Insets;
import javafx.scene.input.TransferMode;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import org.turbomedia.turboedit.editor.events.EventSystem;
import org.turbomedia.turboedit.editor.events.EventType;
import org.turbomedia.turboedit.editor.events.ThemeChangedEventData;
import org.turbomedia.turboedit.editor.misc.StyleManager;
import org.turbomedia.turboedit.editor.windows.errors.FileErrorWindow;
import org.turbomedia.turboedit.shared.project.ProjectFile;

import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;

import static org.turbomedia.turboedit.editor.misc.Locale.GetText;

public class FileBrowserPane extends HBox {

    public FileBrowserPane() throws IOException {
        double GRID_GAP = 6;
        double width = 1050;
        double height = 800;

        var flow = new FlowPane();

        flow.setMinWidth(width);
        flow.setMinHeight(height);

        EventSystem.RegisterListener(EventType.THEME_CHANGED, (dat) -> {
            var data = (ThemeChangedEventData) dat;

            Color color = switch (data.theme()) {
                case DARK -> Color.color(1, 1, 1, 0.05);
                case LIGHT -> Color.color(0, 0, 0, 0.05);
            };

            var background = new Background(new BackgroundFill(color, new CornerRadii(10), Insets.EMPTY));

            flow.setBackground(background);
        });

        StyleManager.CallEvent();

        setOnDragOver((event) -> event.acceptTransferModes(TransferMode.LINK));
        setOnDragDropped((event) -> {
            var dragboard = event.getDragboard();
            var unsupportedFiles = new ArrayList<String>();

            for (var file : dragboard.getFiles()) {
                try {
                    var mimeType = Files.probeContentType(file.toPath());

                    if (!mimeType.startsWith("video/") && !mimeType.startsWith("audio/")) {
                        unsupportedFiles.add(file.getAbsolutePath());
                        continue;
                    }
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }

                System.out.println(file.getAbsoluteFile());
            }

            if (!unsupportedFiles.isEmpty()) {
                new FileErrorWindow(
                        GetText("title.error.file.unsupported"),
                        GetText("error.file.unsupported"),
                        unsupportedFiles
                );
            }
        });

        flow.getChildren().add(new FileCard(new ProjectFile("Test.mp4", "", "video/mp4", null, "", 0, true, 1920, 1080, 60, "AV1", null)));
        flow.setPadding(new Insets(15));
        flow.setVgap(GRID_GAP);
        flow.setHgap(GRID_GAP);

        setPadding(new Insets(15));
        setSpacing(6);

        getChildren().add(new FileActions());
        getChildren().add(flow);
    }
}

File: Editor/src/main/kotlin/turboedit/editor/panes/file_browser/FileBrowserPane.kt
package turboedit.editor.panes.file_browser

import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.scene.input.DragEvent
import javafx.scene.input.TransferMode
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.paint.Paint
import turboedit.editor.events.EventSystem.EventAction
import turboedit.editor.events.EventSystem.registerListener
import turboedit.editor.events.EventType
import turboedit.editor.events.ThemeChangedEventData
import turboedit.editor.misc.Locale
import turboedit.editor.misc.StyleManager
import turboedit.editor.misc.StyleManager.callEvent
import turboedit.editor.windows.errors.FileErrorWindow
import turboedit.shared.project.ProjectFile
import java.io.IOException
import java.nio.file.Files

class FileBrowserPane : HBox() {
    init {
        val gridGap = 6.0
        val width = 1050.0
        val height = 800.0

        val flow = FlowPane()

        flow.minWidth = width
        flow.minHeight = height

        registerListener(EventType.THEME_CHANGED, EventAction { dat: Any? ->
            val data = dat as ThemeChangedEventData?
            val color = when (data!!.theme) {
                StyleManager.Theme.DARK -> Color.color(1.0, 1.0, 1.0, 0.05)
                StyleManager.Theme.LIGHT -> Color.color(0.0, 0.0, 0.0, 0.05)
                else -> {}
            }

            val background = Background(BackgroundFill(color as Paint?, CornerRadii(10.0), Insets.EMPTY))
            flow.background = background
        })

        callEvent()

        onDragOver = EventHandler { event: DragEvent -> event.acceptTransferModes(TransferMode.LINK) }
        onDragDropped = EventHandler { event: DragEvent ->
            val dragboard = event.dragboard
            val unsupportedFiles = ArrayList<String?>()

            for (file in dragboard.files) {
                try {
                    val mimeType = Files.probeContentType(file.toPath())

                    if (!mimeType.startsWith("video/") && !mimeType.startsWith("audio/")) {
                        unsupportedFiles.add(file.absolutePath)
                        continue
                    }
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }

                println(file.getAbsoluteFile())
            }
            if (!unsupportedFiles.isEmpty()) {
                FileErrorWindow(
                    Locale.getText("title.error.file.unsupported"),
                    Locale.getText("error.file.unsupported"),
                    unsupportedFiles
                )
            }
        }

        flow.children
            .add(FileCard(ProjectFile(
                name = "Test.mp4",
                path = "",
                type = "video/mp4",
                previewImage = null,
                hash = "",
                size = 0,
                isVideo = true,
                videoWidth = 1920,
                videoHeight = 1080,
                videoFPS = 60,
                videoCodec = "AV1",
                audio = null
            )))
        flow.padding = Insets(15.0)
        flow.vgap = gridGap
        flow.hgap = gridGap

        padding = Insets(15.0)
        spacing = 6.0

        children.add(FileActions())
        children.add(flow)
    }
}
